lib/cog/repository/groups.ex

```diff
@@ -0,0 +1,215 @@
+defmodule Cog.Repository.Groups do
+  @moduledoc """
+  Behavioral API for interacting with user groups. Prefer these
+  functions over direct manipulation with 'Cog.Repo'.
+  """
+
+  alias Cog.Repo
+  alias Cog.Models.Group
+  alias Cog.Models.User
+  alias Cog.Models.Role
+  import Ecto.Query, only: [from: 1, from: 2]
+
+
+  @doc """
+  Creates a new user group given a map of attributes
+  """
+  @spec new(Map.t) :: {:ok, %Group{}} | {:error, Ecto.Changeset.t}
+  def new(attrs) do
+    %Group{}
+    |> Group.changeset(attrs)
+    |> Repo.insert
+  end
+
+  @doc """
+  Retrieves all user groups
+  """
+  @spec all :: [%Group{}]
+  def all,
+    do: Repo.all(Group) |> Repo.preload([:direct_user_members, :direct_group_members, :roles])
+
+  @doc """
+  Retrieves a single user group based on the id. The given id must be
+  a valid UUID.
+  """
+  @spec by_id(String.t) :: {:ok, %Group{}} | {:error, Ecto.Changeset.t} | {:error, Atom.t}
+  def by_id(id) do
+    if Cog.UUID.is_uuid?(id) do
+      case Repo.get(Group, id) do
+        %Group{} = group ->
+          group = Repo.preload(group, [:direct_user_members, :direct_group_members, :roles])
```

Comment:
Same deal with `direct_group_members` here.

Actually, it might be nice to extract the preloads to a separate private function that we can call in the various places we retrieve groups to easily ensure that we're being consistent.